docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+  'c64': ['c128']
+}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {
+  'i*': [-1.25, 0.5], 'f*': [-0.5, 2], 'c*': [0, 3],
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [1.5, 2], 'f32': [2.5, 2], 'f64': [3.5, 2],
+  'c64': [3, 3], 'c128': [4, 3],
+}
+fig, ax = plt.subplots(figsize=(6, 5))
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+```
+
++++ {"id": "BfHlmmF_GOo_"}
+
+While this does allow for precision-losing promotions between integers and floats, these promotions will not mis-represent the *magnitude* of the result: though the floating point mantissa is not wide enough to represent all values, the exponent is wide enough to approximate them.
+
+This approach also allows a natural promotion path from `int64` to `float64`, though `uint64` remains unpromotable. That said, a connection from `u64` to `f64` would be more defensible here than before.
+
+This promotion scheme still results in some non-ideal promotion paths for accelerator backends; for example operations between `float32` and `uint32` result in `float64`. Additionally, this lattice makes it difficult to find a sensible place to insert `bfloat16` (see below) while maintaining the lattice property.
+
++++ {"id": "7M6EiBDqHNm-"}
+
+### Option 3: Avoid all wider-than-necessary promotions
+
+We can avoid *all* non-ideal 64-bit promotions if we're willing to fundamentally change our thinking around integer and float promotions.
+Just as scalars always defer to the widths of array types, we can make integers always defer to the width of float types:
+
+```{code-cell}
+:cellView: form
+:id: JJ__tn0VJJRD
+:tags: [hide-input]
+
+#@title
+import networkx as nx
+import matplotlib.pyplot as plt
+lattice = {
+  'i*': ['u8', 'i8'], 'f*': ['c*', 'f16'], 'c*': ['c64'],
+  'u8': ['u16', 'i16'], 'u16': ['u32', 'i32'], 'u32': ['u64', 'i64'],
+  'i8': ['i16'], 'i16': ['i32'], 'i32': ['i64'], 'i64': ['f*'],
+  'f16': ['f32'], 'f32': ['f64', 'c64'], 'f64': ['c128'],
+  'c64': ['c128']
+}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {
+  'i*': [-1.25, 0.5], 'f*': [-0.5, 2], 'c*': [0, 3],
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [1.5, 2], 'f32': [2.5, 2], 'f64': [3.5, 2],
+  'c64': [3, 3], 'c128': [4, 3],
+}
+fig, ax = plt.subplots(figsize=(6, 5))
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+```
+
++++ {"id": "xBDCy0AnGsbJ"}
+
+This involves a small slight-of-hand: previously we had used `f*` to refer to a scalar type. In this lattice, `f*` might be applied to the array output of a mixed computation. Instead of thinking of `f*` as a scalar, we could think of it as a special kind of `float64`, with promotion rules that differ from those of `f64`.
```

Comment:
sp. "sleight of hand".

(not "slight").

Reply:
Done.